test/util/TestFeatureRegistry.cpp

```diff
@@ -0,0 +1,132 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+#include <memory>
+
+#include "../../source/FeatureRegistry.h"
+#include "../../source/config/Config.h"
+#include "gtest/gtest.h"
+
+using namespace std;
+using namespace Aws::Iot::DeviceClient;
+using namespace Aws::Iot::DeviceClient::Util;
+
+class FakeFeature : public Feature
+{
+  public:
+    FakeFeature(const std::string &name) : name(name) {}
+    int start() override
+    {
+        started = true;
+        stopped = false;
+        return 0;
+    }
+    int stop() override
+    {
+        stopped = true;
+        started = false;
+        return 0;
+    }
+    std::string getName() { return name; }
+    bool isStarted() { return started; }
+    bool isStopped() { return stopped; }
+
+  private:
+    std::string name;
+    bool started;
+    bool stopped;
+};
+
+class TestFeatureRegistry : public ::testing::Test
+{
+  public:
+    shared_ptr<FeatureRegistry> features;
+    shared_ptr<FakeFeature> feature1;
+    shared_ptr<FakeFeature> feature2;
+    shared_ptr<FakeFeature> feature3;
+
+    void SetUp() override
+    {
+        features = make_shared<FeatureRegistry>();
+        feature1 = make_shared<FakeFeature>("feature-1");
+        feature2 = make_shared<FakeFeature>("feature-2");
+        feature3 = make_shared<FakeFeature>("feature-3");
+    }
+
+    void TearDown() override { features = make_shared<FeatureRegistry>(); }
+};
+
+TEST_F(TestFeatureRegistry, AddFeaturesToRegistry)
+{
+    features->add(feature1->getName(), feature1);
+    ASSERT_EQ(1, features->getSize());
+
+    features->add(feature1->getName(), feature1);
+    ASSERT_EQ(1, features->getSize());
+
+    features->add(feature2->getName(), feature2);
+    ASSERT_EQ(2, features->getSize());
+
+    features->add(feature3->getName(), feature3);
+    ASSERT_EQ(3, features->getSize());
+}
+
+TEST_F(TestFeatureRegistry, GetNonExistentFeature)
+{
+    shared_ptr<Feature> feature = features->get("Not a Feature");
+    ASSERT_EQ(nullptr, feature);
+}
+
+TEST_F(TestFeatureRegistry, GetFeatureByName)
+{
+    features->add(feature1->getName(), feature1);
+    shared_ptr<Feature> feature = features->get(feature1.get()->getName());
+    ASSERT_EQ(feature1, feature);
+}
+
+TEST_F(TestFeatureRegistry, TestRemoveFeature)
+{
+    features->add(feature1->getName(), feature1);
+    features->add(feature2->getName(), feature2);
+    ASSERT_EQ(2, features->getSize());
+    features->remove(feature1->getName());
+    ASSERT_EQ(nullptr, features->get(feature1->getName()));
+    features->remove(feature2->getName());
+    ASSERT_EQ(nullptr, features->get(feature2->getName()));
+}
+
+TEST_F(TestFeatureRegistry, StartAllFeatures)
+{
+    features->add(feature1->getName(), feature1);
+    features->add(feature2->getName(), feature2);
+    features->add(feature3->getName(), feature3);
+
+    features->startAll();
+
+    ASSERT_TRUE(feature1->isStarted());
```

Comment:
I think there is an ASSERT_ALL where you could group these together to be more readable.

Reply:
Not sure if I see it on here: https://google.github.io/googletest/reference/assertions.html